symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Optimization.hs

```diff
@@ -0,0 +1,91 @@
+import           Data.Bool                                               (bool)
+import           Data.ByteString                                         (ByteString)
+import           Data.Functor                                            ((<&>))
+import           Data.Functor.Rep                                        (Representable (..))
+import           Data.Map                                                hiding (drop, foldl, foldr, map, null, splitAt,
+                                                                          take)
+import qualified Data.Map.Internal                                       as M
+import qualified Data.Map.Monoidal                                       as MM
+import qualified Data.Set                                                as S
+import           Prelude                                                 hiding (Num (..), drop, length, product,
+                                                                          splitAt, sum, take, (!!), (^))
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Basic.Number
+import           ZkFold.Base.Algebra.Polynomials.Multivariate            (evalMonomial)
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Monomial   (Mono (..), oneM)
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Polynomial (Poly (..), evalPolynomial, var)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Instance     ()
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Witness      (WitnessF (..))
+
+--------------------------------- High-level functions --------------------------------
+
+-- | Replaces linear polynoms of the form
+-- @(fromConstant k) * (NewVar nV) + (fromConstant c)@
+-- with a constant variable @fromConstant $ negate c // k@ in an arithmetic circuit
+-- and replaces variable with a constant in witness
+--
+optimize :: forall a p i o.
+  (Arithmetic a, Ord (Rep i), Functor o, Binary (Rep i), Binary a, Binary (Rep p)) =>
+  ArithmeticCircuit a p i o -> ArithmeticCircuit a p i o
+optimize (ArithmeticCircuit s r w o) = ArithmeticCircuit {
+    acSystem = addInVarConstraints newS,
+    acRange = optRanges vs r,
+    acWitness = (>>= optWitVar vs) <$>  M.filterWithKey (\k _ -> notMember (NewVar k) vs) w,
+    acOutput = o <&> \case
+      SysVar sV -> maybe (SysVar sV) ConstVar (M.lookup sV vs)
+      so -> so}
+  where
+    (newS, vs) = varsToReplace (s, M.empty)
+
+    addInVarConstraints :: Map ByteString (Poly a (SysVar i) Natural) -> Map ByteString (Poly a (SysVar i) Natural)
+    addInVarConstraints p = p <> fromList [(polyId, poly) | (InVar i, v) <- assocs vs,
+                                                            let poly = var (InVar i) - fromConstant v,
+                                                            let polyId = toVar @a @p @i (pure $ WSysVar (InVar i))]
```

Comment:
In `Internal`, constraint id is computed on the whole constraint, not on the variable!